Batteries/Data/Vector/Basic.lean

```diff
@@ -0,0 +1,380 @@
+-/
+def setN (v : Vector α n) (i : Nat) (h : i < n := by get_elem_tactic) (x : α) : Vector α n :=
+  ⟨v.toArray.set (Fin.cast v.size_eq.symm ⟨i, h⟩) x, proof⟩
+  where
+    proof := by {rw [size_set]; exact v.size_eq}
+/--
+Sets an element in a vector, or do nothing if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def setD (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.setD i x, proof⟩
+  where
+    proof := by
+      rw [size_setD]
+      exact v.size_eq
+
+/--
+Sets an element in an array, or panic if the index is out of bounds.
+
+This will perform the update destructively provided that a has a reference count of 1 when called.
+-/
+def set! (v : Vector α n) (i : Nat) (x : α) : Vector α n :=
+  ⟨v.toArray.set! i x, proof⟩
+  where
+    proof := by
+      rw [set!_is_setD, size_setD]
+      exact v.size_eq
+
+/-- Appends a vector to another. -/
+def append : Vector α n → Vector α m → Vector α (n + m)
+  | ⟨a₁, h₁⟩, ⟨a₂, h₂⟩ => ⟨a₁ ++ a₂, by rw[Array.size_append, ←h₁,←h₂]⟩
+
+instance : HAppend (Vector α n) (Vector α m) (Vector α (n + m)) where
+  hAppend := append
+
+/-- Creates a vector from another with a provably equal length. -/
+protected def cast {n m : Nat} (h : n = m) : Vector α n → Vector α m
+  | ⟨x, p⟩ => ⟨x, h ▸ p⟩
+
+/--
+`extract v start halt` Returns the slice of `v` from indices `start` to `stop` (exclusive).
+If `start` is greater or equal to `stop`, the result is empty.
+If `stop` is greater than the size of `v`, the size is used instead.
+-/
+def extract (v : Vector α n) (start stop : Nat) : Vector α (min stop n - start) :=
+  ⟨Array.extract v.toArray start stop, proof⟩
+  where
+    proof := by
+      rw [size_extract, v.size_eq]
+
+/-- Custom eliminator for `Vector α n` through Arrays -/
+@[elab_as_elim]
+def elimAsArray {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : Array α, motive ⟨a, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨a, rfl⟩ => mk a
+
+/-- Custom eliminator for `Vector α n` throughLists -/
+@[elab_as_elim]
+def elimAsList {motive : ∀ {n}, Vector α n → Sort u} (mk : ∀ a : List α, motive ⟨⟨a⟩, rfl⟩) :
+    {n : Nat} → (v : Vector α n) → motive v
+  | _, ⟨⟨a⟩, rfl⟩ => mk a
+
+/-- Maps a vector under a function. -/
+def map (f : α → β) : Vector α n → Vector β n
+  | ⟨a, h⟩ => ⟨Array.map f a, by simp [*]⟩
+
+/-- Maps two vectors under a curried function of two variables. -/
+def zipWith : Vector α n → Vector β n → (α → β → φ) → Vector φ n
+  | ⟨a, h₁⟩, ⟨b, h₂⟩, f => ⟨Array.zipWith a b f, by simp [Array.size_zipWith, h₁, h₂]⟩
+
+/-- Drops `i` elements from a vector of length `n`; we can have `i > n`. -/
+def drop (i : Nat) (v : Vector α n) : Vector α (n - i) :=
+  have : min n n - i = n - i := by
+    rw [Nat.min_self]
+  Vector.cast this (extract v i n)
+
+/-- Takes `i` elements from a vector of length `n`; we can have `i > n`. -/
+alias take := shrink
```

Comment:
`Vector.shrink` needs to be defined before this. Right now this is an alias for `Array.shrink`.

Reply:
Done.